include/nbl/ext/FFT/FFT.h

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_EXT_FFT_INCLUDED_
+#define _NBL_EXT_FFT_INCLUDED_
+
+#include "nabla.h"
+#include "nbl/video/IGPUShader.h"
+#include "nbl/asset/ICPUShader.h"
+
+namespace nbl
+{
+namespace ext
+{
+namespace FFT
+{
+
+class FFT : public core::TotalInterface
+{
+	public:
+
+		enum class Direction : uint32_t {
+			X = 0,
+			Y = 1,
+			Z = 2,
+		};
+		
+		enum class PaddingType : uint32_t {
+			CLAMP_TO_EDGE = 0,
+			FILL_WITH_ZERO = 1,
+		};
+
+		enum class DataType {
+			SSBO,
+			TEXTURE2D,
+		};
+
+		struct DispatchInfo_t
+		{
+			uint32_t workGroupDims[3];
+			uint32_t workGroupCount[3];
+		};
+
+		struct alignas(16) Uniforms_t 
+		{
+			uint32_t dims[3];
+		};
+
+		// returns dispatch size and fills the uniform data
+		static inline DispatchInfo_t buildParameters(
+			asset::VkExtent3D const & paddedInputDimensions,
+			Direction direction,
+			uint32_t num_channels)
+		{
+			assert(num_channels > 0);
+			assert(core::isPoT(paddedInputDimensions.width) && core::isPoT(paddedInputDimensions.height) && core::isPoT(paddedInputDimensions.depth));
+			DispatchInfo_t ret = {};
+
+			ret.workGroupDims[0] = DEFAULT_WORK_GROUP_X_DIM;
+			ret.workGroupDims[1] = 1;
+			ret.workGroupDims[2] = 1;
+
+			if(direction == Direction::X)
+			{
+				ret.workGroupCount[0] = num_channels;
+				ret.workGroupCount[1] = paddedInputDimensions.height;
+				ret.workGroupCount[2] = paddedInputDimensions.depth;
+			}
+			else if(direction == Direction::Y) {
+				ret.workGroupCount[0] = paddedInputDimensions.width;
+				ret.workGroupCount[1] = num_channels;
+				ret.workGroupCount[2] = paddedInputDimensions.depth;
+			}
+			else if(direction == Direction::Z)
+			{
+				ret.workGroupCount[0] = paddedInputDimensions.width;
+				ret.workGroupCount[1] = paddedInputDimensions.height;
+				ret.workGroupCount[2] = num_channels;
+			}
+
+			return ret;
+		}
+
+		//
+		static core::SRange<const asset::SPushConstantRange> getDefaultPushConstantRanges();
+
+		//
+		static core::SRange<const video::IGPUDescriptorSetLayout::SBinding> getDefaultBindings(video::IVideoDriver* driver, DataType inputType);
+		
+		//
+		static inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultPipelineLayout(video::IVideoDriver* driver, DataType inputType)
+		{
+			auto pcRange = getDefaultPushConstantRanges();
+			auto bindings = getDefaultBindings(driver, inputType);
+			return driver->createGPUPipelineLayout(
+				pcRange.begin(),pcRange.end(),
+				driver->createGPUDescriptorSetLayout(bindings.begin(),bindings.end()),nullptr,nullptr,nullptr
+			);
+		}
+		
+		//
+		static inline size_t getOutputBufferSize(asset::VkExtent3D const & paddedInputDimensions, uint32_t numChannels)
```

Comment:
again, just pad the dimensions for the user.

Reply:
@devshgraphicsprogramming 
There are 2 choices for me.

Either get the other images size as an input and do 
`padToNextPOT(max(InputDimensions, otherImageSize));`
or
The user has to give me the maximum of the 2 dimensions (kernel and srcImage) to Pad.

Since these 2 choices seemed misleading to me, I initially decided to just get the `paddedDimensions` as the input.

Anyways. just let me know what to do and I'll do it. 

I can also move the `padDimensionsToNextPOT` function from main.cpp to FFT Class.